File: numerical_methods/root_finding/newton_raphson/newton_raphson_2.js
/*
Finding the roots of functions is important in many engineering
applications such as signal processing and optimization.
*/
const f = (x) => x**3 - 100*x**2 - x + 100;
const df = (x) => 3*x**2 - 200*x - 1;

function newton_raphson(f, df, x0, tol) {
  if (Math.abs(f(x0)) < tol) {
    return x0;
  }
  else {
    return newton_raphson(f, df, x0 - f(x0)/df(x0), tol);
  }
}

//const root = newton_raphson(f, df, 0, 1e-6);
// root = 100

//const root = newton_raphson(f, df, 0.005, 1e-6);
// root = 1.00000

const root = newton_raphson(f, df, -0.005, 1e-6);
// root = -1.00000

console.log("root:", root);

const replacing = f(root);
console.log("replacing:", replacing);

File: numerical_methods/root_finding/newton_raphson/newton_raphson_2.py
#!/usr/bin/env python3
f = lambda x : x**3 - 100*x**2 - x + 100
df = lambda x : 3*x**2 - 200*x - 1

def newton_raphson(f, df, x0, tol):
    if abs(f(x0)) < tol:
        return x0
    else:
        return newton_raphson(f, df, x0 - f(x0)/df(x0), tol)


# root = newton_raphson(f, df, 0, 1e-6)
# root = 100.0

# root = newton_raphson(f, df, 0.005, 1e-6)
# root = 1.00000

root = newton_raphson(f, df, -0.005, 1e-6)
# root = -1.00000

print("root:", root)

replacing = f(root)

print("replacing:", replacing)
